benchmark/array/array_benchmark.cpp

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Team Dissolve and contributors
+#include "benchmark/benchmark.h"
+#include "templates/array.h"
+#include "templates/array2d.h"
+#include "templates/array3d.h"
+
+#include <numeric>
+#include <random>
+
+// Benchmark array
+
+// Benchmarks random access into the array structure
+
+static void BM_Array_1d(benchmark::State &state)
+{
+    int bytes = state.range(0);
+    int numVals = bytes / sizeof(int) / 2;
+    std::default_random_engine generator;
+    std::uniform_int_distribution<int> distribution(std::numeric_limits<int>::min(), std::numeric_limits<int>::max());
+    std::vector<int> indicies(numVals);
+    std::iota(indicies.begin(), indicies.end(), 0);
+    std::shuffle(indicies.begin(), indicies.end(), generator);
+    Array<int> array(numVals);
+    for (int i = 0; i < array.size(); ++i)
+    {
+        array[i] = distribution(generator);
+    }
+
+    for (auto _ : state)
+    {
+        double sum = 0;
+        for (int i : indicies)
+            sum += array[i];
+        benchmark::DoNotOptimize(sum);
+    }
+    state.SetBytesProcessed(long(state.iterations()) * long(bytes));
+    state.SetLabel(std::to_string(bytes / 1024) + "kb");
+}
+
+static void BM_Array_2d(benchmark::State &state)
+{
+    int bytes = state.range(0);
+    int numVals = (bytes / sizeof(int)) / 2;
+    std::default_random_engine generator;
+    std::uniform_int_distribution<int> distribution(std::numeric_limits<int>::min(), std::numeric_limits<int>::max());
+    int dim = sqrt(numVals);
+    std::vector<int> indicies(dim * dim);
+    std::iota(indicies.begin(), indicies.end(), 0);
+    std::shuffle(indicies.begin(), indicies.end(), generator);
+
+    Array2D<int> array(dim, dim);
+    for (int i = 0; i < array.nRows(); ++i)
+        for (int j = 0; j < array.nColumns(); ++j)
+            array[{i, j}] = distribution(generator);
```

Comment:
```suggestion
    std::generate(array.begin(), array.end(), [&](){return distriburion(generator);}); 
```

Went with `generate` here instead of `fill` because I don't know which one we're supposed to use here.  Also just somewhat showing off that you can iterate over the entire Array2D.

Reply:
I assume that this is the correct intent (note typo in `distriburion` in the suggestion).